src/cachingreader.cpp

```diff
@@ -123,18 +123,6 @@ CachingReaderChunkForOwner* CachingReader::allocateChunk(SINT chunkIndex) {
     //qDebug() << "Allocating chunk" << pChunk << pChunk->getIndex();
     m_allocatedCachingReaderChunks.insert(chunkIndex, pChunk);
 
-    // Adjust the least-recently-used item before inserting the
-    // chunk as the new most-recently-used item.
-    if (nullptr == m_lruCachingReaderChunk) {
-        if (nullptr == m_mruCachingReaderChunk) {
-            m_lruCachingReaderChunk = pChunk;
-        } else {
-            m_lruCachingReaderChunk = m_mruCachingReaderChunk;
-        }
-    }
-    // Insert the chunk as the new most-recently-used item.
-    pChunk->insertIntoListBefore(m_mruCachingReaderChunk);
```

Comment:
Chunks need to be inserted into the list after retaining ownership back from 
the worker. Thanks for asking, Daniel! Looks look like I missed one case, 
should be fixed now.

Nevertheless the design is still very fragile, because it violates basic OO 
encapsulation principles.

On 04/05/2016 12:36 AM, Daniel Schürmann wrote:
>
> In src/cachingreader.cpp 
> <https://github.com/mixxxdj/mixxx/pull/914#discussion_r58460897>:
>
> > @@ -123,18 +123,6 @@ CachingReaderChunkForOwner* CachingReader::allocateChunk(SINT chunkIndex) {
> >      //qDebug() << "Allocating chunk" << pChunk << pChunk->getIndex();
> >      m_allocatedCachingReaderChunks.insert(chunkIndex, pChunk);
> >
> > -    // Adjust the least-recently-used item before inserting the
> > -    // chunk as the new most-recently-used item.
> > -    if (nullptr == m_lruCachingReaderChunk) {
> > -        if (nullptr == m_mruCachingReaderChunk) {
> > -            m_lruCachingReaderChunk = pChunk;
> > -        } else {
> > -            m_lruCachingReaderChunk = m_mruCachingReaderChunk;
> > -        }
> > -    }
> > -    // Insert the chunk as the new most-recently-used item.
> > -    pChunk->insertIntoListBefore(m_mruCachingReaderChunk);
>
> I did not got the point why it is OK, not to add the new chunk to the 
> linked list.
> Is it sure, that the chunk ends finally in th list?
> I am sure it is, but I don't see it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/914/files/95f52e27b90b726c7457097118d730ef3a67027a#r58460897>
>